pkg/eval/str/str.go

```diff
@@ -92,6 +95,74 @@ import (
 // ▶ $false
 // ```
 
+//elvdoc:fn from-codepoints
+//
+// ```elvish
+// str:from-codepoints $number...
+// ```
+//
+// Outputs a string consisting of the given Unicode codepoints. Example:
+//
+// ```elvish-transcript
+// ~> str:from-codepoints 0x61
+// ▶ a
+// ~> str:from-codepoints 0x4f60 0x597d
+// ▶ 你好
+// ```
+//
+// @cf str:to-codepoints
+
+func fromCodepoints(nums ...int) (string, error) {
+	var b bytes.Buffer
+	for _, num := range nums {
+		if int(rune(num)) != num {
+			return "", errs.OutOfRange{
+				What:     "codepoint",
+				ValidLow: -2147483648, ValidHigh: 2147483647,
+				Actual: strconv.Itoa(num)}
+		}
+		if !utf8.ValidRune(rune(num)) {
+			return "", fmt.Errorf("invalid codepoint: %d", num)
+		}
+		b.WriteRune(rune(num))
+	}
+	return b.String(), nil
+}
+
+//elvdoc:fn from-utf8-bytes
+//
+// ```elvish
+// str:from-from-utf8-bytes $number...
+// ```
+//
+// Outputs a string consisting of the given Unicode bytes. Example:
+//
+// ```elvish-transcript
+// ~> str:from-utf8-bytes 0x61
+// ▶ a
+// ~> str:from-utf8-bytes 0xe4 0xbd 0xa0 0xe5 0xa5 0xbd
+// ▶ 你好
+// ```
+//
+// @cf str:to-utf8-bytes
+
+func fromUtf8Bytes(nums ...int) (string, error) {
+	var b bytes.Buffer
+	for _, num := range nums {
+		if int(byte(num)) != num {
+			return "", errs.OutOfRange{
+				What:     "byte",
+				ValidLow: 0, ValidHigh: 255,
+				Actual: strconv.Itoa(num)}
+		}
+		b.WriteByte(byte(num))
+	}
+	if !utf8.Valid(b.Bytes()) {
+		return "", fmt.Errorf("invalid utf8 bytes: %d", b.Bytes())
```

Comment:
Return a BadValue struct here. You can use "arguments to str:from-utf8-bytes" as What and "valid UTF-8 sequence" as Valid.